### ISSUE TITLE
Support supplemental parameters when used with .bicepparam parameter file not working

### ISSUE THREAD

**Issue by tnsholding** [Azure/azure-cli#27924]
### Describe the bug

Trying to use the new possibility for using supplemental parameters when used with .bicepparam files according to https://github.com/Azure/azure-cli/pull/27527.

This change is supposed to be part of az cli version 2.54.0.

But running this command in an Azure DevOps pipeline does not seem to work when passing in multiple inline parameters. 
Trying to run the following code in an AzureCLI@2 task:

```
 az deployment group create 
  --name api-deleteme-api 
  --subscription '62394fa1-c454-4478-944e-3bcadc25b8d1' 
  --resource-group 'integrationServices-dev' 
  --template-file 'D:\a\1/drop/api.bicep' 
  --parameters 'D:\a\1/drop/deleteme-api/azuredeploy.parameters.bicepparam' 
  --parameters version='v0.45' apiRevision='13' apiManagementServiceName='hofor-api-management-dev' 
```
Having a bicepparam file with the following content 

```
...other parameters
param apiManagementServiceName = '#apiManagementServiceName - override from devops pipeline#' 

```

When running the deployment I get the error:
_The Resource 'Microsoft.ApiManagement/service/#apiManagementServiceName - override from devops pipeline#' under resource group 'integrationServices-dev' was not found_

Clearly indicating that the parameter value is still the value from the bicepparam file and NOT being overridden by the inline parameter passed to az deployment command. 

Is this not how it was supposed to work? - the above code is running in the AzureCLI@2 task with the correct version (2.54.0):

![image](https://github.com/Azure/azure-cli/assets/4534590/875aa93f-0289-4ff0-9d79-36576684b8e7)

### Related command

 az deployment group create 
  --name api-deleteme-api 
  --subscription '62394fa1-c454-4478-944e-3bcadc25b8d1' 
  --resource-group 'integrationServices-dev' 
  --template-file 'D:\a\1/drop/api.bicep' 
  --parameters 'D:\a\1/drop/deleteme-api/azuredeploy.parameters.bicepparam' 
  --parameters version='v0.45' apiRevision='13' apiManagementServiceName='hofor-api-management-dev' 

### Errors

The Resource 'Microsoft.ApiManagement/service/#apiManagementServiceName - override from devops pipeline#' under resource group 'integrationServices-dev' was not found

### Issue script & Debug output

The Resource 'Microsoft.ApiManagement/service/#apiManagementServiceName - override from devops pipeline#' under resource group 'integrationServices-dev' was not found

### Expected behavior

Expected the inline paramater to override the parameter in the bicepparam file. 

### Environment Summary

  "azure-cli": "2.54.0",
  "azure-cli-core": "2.54.0",

**Comment by yonzhan** [comment]
Thank you for opening this issue, we will look into it.

**Comment by anthony-c-martin** [comment]
@tnsholding - could you run the following command before starting the deployment and share the output, just to confirm the version of Bicep that has been installed:
```sh
az bicep version
```

I've tried to repro this locally with a minimal repro, and have not been able to:
Command:
```sh
az deployment group create -g ant-test --parameters main.bicepparam --query "properties.outputs" --parameters foo='override foo' bar='override bar'
```
Output:
```json
{
  "bar": {
    "type": "String",
    "value": "override bar"
  },
  "baz": {
    "type": "String",
    "value": "base baz"
  },
  "foo": {
    "type": "String",
    "value": "override foo"
  }
}
```
main.bicepparam:
```bicep
using 'main.bicep'

param foo = 'base foo'
param bar = 'base bar'
param baz = 'base baz'
```
main.bicep:
```bicep
param foo string
param bar string
param baz string

output foo string = foo
output bar string = bar
output baz string = baz
```

**Comment by tnsholding** [comment]
Adding following to output just before running  az deployment group create:

`az bicep version`

returns:

Bicep CLI version 0.23.1

**Comment by tnsholding** [comment]
It was actually a variable not being set correctly in my pipeline, so it appears to be working now.

However one thing seems to be different from json parameter files:
I need to include the parameters in the bicepparam file in order to be able to override them. I would like the possibility to have them defined in the bicep file with default values, NOT including them in the bicepparam file, but being able to override default values with inline parameters. 

This does not seem to be possible? - if not defined in bicepparam, the inline parameters are not being used.

**Comment by cwoz117** [comment]
Interesting to note:

When my bicep file has @allowed, it errors out before it reads the overridden env variable:

#Bicep param file
`param env = '# to be overridden`

# Bicep File
```
@allowed([
    'dev'
    'test'
    'prod'
])
param env string
```

# az cli to deploy:
```
az group create
    -g $(rg) \
    -p main.bicepparam \
    -p env='dev'
```

Without the `@allowed` decorator this works just fine

**Comment by anthony-c-martin** [comment]
@cwoz117 thanks - that helped pinpoint the issue!

I believe the problem is that we attempt to compile the param file first without the overrides [here](https://github.com/Azure/azure-cli/blob/a9dc0fd41e5afd10765feaaaccfebcb7e11f0501/src/azure-cli/azure/cli/command_modules/resource/custom.py#L1127), which throws an exception before we actually attempt to compile with the overrides [here](https://github.com/Azure/azure-cli/blob/a9dc0fd41e5afd10765feaaaccfebcb7e11f0501/src/azure-cli/azure/cli/command_modules/resource/custom.py#L1139). This looks like a genuine bug in the command.

**Comment by torreymicrosoft** [comment]
Thanks for your patience — this was recently picked up as an outstanding open issue for the ARM Deployments team to review.

**On the original issue (@tnsholding):** This appears to have been resolved — a pipeline environment variable wasn't being passed correctly. Separately, you raised a good design question: parameters not defined in a `.bicepparam` file can't be overridden via inline `--parameters`. This is by design — the `.bicepparam` file is compiled by Bicep first, and it must include any parameter you intend to override inline.

**On the secondary bug reported by @cwoz117:** I've reproduced this on CLI **2.85.0** with Bicep **0.43.8** — it seems the fix was never filed, as the issue is still present. When a `.bicepparam` file contains a placeholder value that fails an `@allowed` constraint, the CLI errors out with BCP033 even when a valid override is passed inline:

```
ERROR: BCP033: Expected a value of type "'dev' | 'prod' | 'test'" but the provided value is of type "'# to be overridden'"
```

Root cause (first identified by @anthony-c-martin in Dec 2023): `_parse_bicepparam_file` in `custom.py` compiles the `.bicepparam` file **without** the inline overrides first to extract the template reference. If the placeholder value violates `@allowed`, compilation fails here — before the second compile that would apply your overrides.

**Workaround (verified):** Use a valid `@allowed` value as your placeholder instead of a non-conforming string:

```bicep
// main.bicepparam
using 'main.bicep'

param env = 'dev'  // valid placeholder; will be overridden by --parameters env=prod at deploy time
```

This secondary bug should be tracked separately. Please consider filing a new issue titled something like: `az deployment group create: BCP033 on @allowed constraint when using inline --parameters override with .bicepparam placeholder` and referencing this thread + @anthony-c-martin's Dec 2023 diagnosis.

Closing this issue as the original question was answered. Thanks to everyone who contributed to the thread.
